src/main/kotlin/org/opensearch/replication/repository/RestoreContext.kt

```diff
@@ -12,47 +12,47 @@
 package org.opensearch.replication.repository
 
 import org.apache.lucene.index.IndexCommit
-import org.opensearch.replication.util.performOp
 import org.apache.lucene.store.IOContext
 import org.apache.lucene.store.IndexInput
-import org.opensearch.OpenSearchException
 import org.opensearch.common.concurrent.GatedCloseable
-import org.opensearch.index.engine.Engine
 import org.opensearch.index.shard.IndexShard
 import org.opensearch.index.store.Store
 import java.io.Closeable
+import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.locks.ReentrantLock
+import kotlin.concurrent.withLock
 
-class RestoreContext(val restoreUUID: String,
-                     val shard: IndexShard,
-                     val indexCommitRef: GatedCloseable<IndexCommit>,
-                     val metadataSnapshot: Store.MetadataSnapshot,
-                     val replayOperationsFrom: Long): Closeable {
+class RestoreContext(
+    val restoreUUID: String,
+    val shard: IndexShard,
+    val indexCommitRef: GatedCloseable<IndexCommit>,
+    val metadataSnapshot: Store.MetadataSnapshot,
+    val replayOperationsFrom: Long,
+) : Closeable {
 
-    companion object {
-        private const val INITIAL_FILE_CACHE_CAPACITY = 20
-    }
-    private val currentFiles = LinkedHashMap<String, IndexInput>(INITIAL_FILE_CACHE_CAPACITY)
+    private val currentFiles = ConcurrentHashMap<String, IndexInput>(INITIAL_FILE_CACHE_CAPACITY)
+
+    private val fileLocks = ConcurrentHashMap<String, ReentrantLock>()
 
     fun openInput(store: Store, fileName: String): IndexInput {
-        var currentIndexInput = currentFiles.getOrDefault(fileName, null)
-        if(currentIndexInput != null) {
-            return currentIndexInput.clone()
-        }
-        store.performOp({
-            currentIndexInput = store.directory().openInput(fileName, IOContext.READONCE)
-        })
+        val lock = fileLocks.computeIfAbsent(fileName) { ReentrantLock() }
 
-        currentFiles[fileName] = currentIndexInput!!
-        return currentIndexInput!!.clone()
+        lock.withLock {
+            val baseInput = currentFiles.computeIfAbsent(fileName) {
+                store.directory().openInput(fileName, IOContext.DEFAULT)
```

Comment:
https://github.com/apache/lucene/blob/219cd5eaebca686c1df93226f47924337ee1a783/lucene/MIGRATE.md#iocontextload-and-iocontextread-removed

`IOContext.READ` has been removed, it should be replaced with `IOContext.DEFAULT`.

Reply:
By changing this option, I expect that `MemorySegmentIndexInput` will no longer be used, and instead, a combination of `BufferedIndexInput` and `MMapIndexInput` will be used.
Although this may result in slightly reduced performance for random access reads, it seems to be an unavoidable trade-off given the current multi-threaded architecture of CCR.

For now, we should prioritize fixing the bug and then explore ways to further improve overall index read performance by optimizing the CCR architecture.

Reply:
Just an observation - `store.directory().openInput()` used to be within the`store.performOp` call (i.e. within `Store.incRef`/`Store.decRef`) and now it's not.

Reply:
@Mishail 

> Just an observation - store.directory().openInput() used to be within thestore.performOp call (i.e. within Store.incRef/Store.decRef) and now it's not.

Thank you! I completely missed that, I’ll make the fix soon.